### PR TITLE
feat: support multi-line address text (for pdf generation)

### LINKED ISSUE #3062: support multi-line address text (for pdf generation
For printed letters, users would like to include the address text in PDF template files in a way that it is output in multiple lines:
> Street, No
> zip code, City

### PR DESCRIPTION
**Pull request by sadaf895**
- closes #3062

### PR Checklist

_Before you request a manual PR review from someone, please go through all of this list, check the points and mark them checked here.
This helps us reduce the number of review cycles and use the time of our core reviewers more effectively._

- [ ] all automatic CI checks pass (🟢)
  - i.e. code formatting (`ng lint`) passes
  - i.e. unit tests (`ng test`) passes
  - in some cases unit test coverage can be difficult to achieve. If that or any other CI check fails, you can also leave a short comment here why this should be accepted anyway.
- [ ] manually tested (all) required functionality described in the issue manually yourself on the final version of the code (developer)
- [ ] reviewed the "Files changed" section of the PR briefly yourself to check for any unwanted changes
  - e.g. clean up debugging console.log statements, disabled tests, ...
  - please also avoid changes that are not directly related to the issue of the PR, even small code reformatting changes make the review process much more complex
- [ ] 🚦 PR status is changed to "Ready for Review"
  - while you are still working on initial implementation, keep the PR in "Draft" status
  - once you are done with your initial work, change to "Ready for Review". This will trigger some additional automated checks.
  - (PR = "Ready for Review" does not immediately request a manual review yet, first complete the further checks below)
- [ ] add label **"Final Review"** to trigger Argos visual review and CodeRabbit AI review (can also be done while still in Draft)
- [ ] marked each code review comment as resolved OR commented on it with a question, if unsure
  - both implementing suggestions of automatic code review or discarding them as not applicable is okay
- [ ] all checkboxes in this checklist are checked (to show the reviewer this really is ready)
- [ ] 🚦 moved the issue related to this PR to Status "Functional Review" to request a personal review

⏪ if issues are commented from testing or code review, make changes. Please then re-check every item of the checklist here again.

---

### Remaining TODOs:

- [ ] unit tests

## Summary by CodeRabbit

* **Bug Fixes**
  * Preserve top-level address fields (road, house number, postcode, city, country) when switching between lookup, manual edits, or map selections.
  * Replace stale address parts on location changes and use village as a fallback when city/town is missing.

* **New Features**
  * Enriched lookup results with consistent address parts and postcode normalized to string for reliable display and PDF output.

### CODE REVIEW

**Comment by sadaf895** [comment]
# Manual testing checklist — Template export & address fields

Purpose
- Verify top-level address fields (road, house_number, postcode, city, country) are populated, preserved through UI flows, and available to template generation.

Quick smoke test:
1. Open an entity with a `location` field (e.g., a child or school record).
2. Open the location editor (`EditLocation`) and search for an address (e.g., "Main St 42 Berlin").
3. Select a suggestion. Verify the location string shows multi-line address text.
4. Inspect the saved object (via browser devtools or network) and confirm `geoLookup` contains `road`, `house_number`, `postcode`, `city`, `country` and that the top-level `location` object also exposes these fields (or `location.geoLookup.*`).

Template export flow
1. On an entity record, open the entity actions menu and choose "Generate File".
2. In the Template Export dialog, select a template and click to generate.
3. If backend is reachable and returns a PDF, verify download and that the generated document contains the expected address fields (where template placeholders used). If backend not reachable, verify the dialog does not crash and shows an informative message.

Edge cases to test
- City fallback: search/select an address where `city` is missing but `village` or `town` exist; confirm `city` is populated from `village`/`town`.
- Postcode types: test addresses where `postcode` is numeric and where it is string — ensure templates receive a string value.
- Missing address parts: select addresses without house number or road and confirm display_name and saved fields behave sensibly.
- Manual edit: in Address editor, change the manual multi-line text and save; verify previously-populated top-level fields are preserved when appropriate, and removed if the manual text is cleared.

What to look for (expected behavior)
- `geoLookup` contains populated `road`, `house_number`, `postcode` (string), `city`, `country` when available.
- `city` falls back to `village` or `town` if `city` missing.
- UI flows (address search, map selection, manual edit) preserve enriched fields when applicable.
- Template generation passes the entity (including `location` with the new fields) to the export API without throwing errors.

**Comment by Abhinegi2** [comment]
Functionality is working as expected. For testing, I also added a new template on the dev site: "Address pr test". (doing code review meanwhile)

One UI/UX question: Are these buttons already intended to look like this?(Edit Address / Clear)
image

These buttons feel a bit inconsistent / slightly weird visually

File: src/app/features/location/address-edit/address-edit.component.ts
Comment (Abhinegi2) [inline src/app/features/location/address-edit/address-edit.component.ts:73]: Both updateLocationString and updateFromAddressSearch manually copy each of the 5 fields into the object passed to updateLocation. But updateLocation already runs enrichGeoLocation, which would fill those fields from geoLookup automatically?

File: src/app/features/location/geo.service.ts [hunk 1]
```diff
@@ -24,6 +24,12 @@ import {
 
 export interface GeoResult extends Coordinates {
   display_name: string;
+  // top-level, commonly-used address parts populated from `address` when available
+  road?: string;
+  house_number?: string;
+  postcode?: string;
```
Comment (Abhinegi2) [inline src/app/features/location/geo.service.ts]: I noticed that in some other places postcode?: number | string, but in some cases it's only string.

Also, should house_number support number as well, or should it remain only string?

**Comment by sleidig** [comment]
> One UI/UX question: Are these buttons already intended to look like this?(Edit Address / Clear) image
> 
> These buttons feel a bit inconsistent / slightly weird visually

That has been the UI for this for a while. I think putting these actions into the small hint section below the field was done to save space (as these options are really an edge case and shouldn't be used commonly). UX is not great of this (the "edit address text" doesn't change once you are in edit mode, for example) but I guess optimizing this is a separate topic to consider. (Thanks for bringing it up, though! :handshake: )

File: src/app/features/location/geo.service.ts (hunk 1 above)
Comment (sleidig) [reply src/app/features/location/geo.service.ts]: I think better if we keep these as string only. We might have issues with leading zeroes and also may need letter suffixes, so best always convert that to string in my opinion.

**Comment by Abhinegi2** [comment]
also, sometimes when you first type an address and save it, the correct information is saved. But later, if you select a different location on the map from the previously selected city, duplicate/conflicting address data gets saved in the DB.

```json
"address": {
  "locationString": "Berliner Straße, 16727 Marwitz",
  "geoLookup": {
    "address": {
      "road": "Berliner Straße",
      "municipality": "Oberkrämer",
      "county": "Oberhavel",
      "state": "Brandenburg",
      "postcode": "16727",
      "country": "Germany"
    },
    "road": "Berliner Straße",
    "postcode": "16727",
    "city": "Marwitz",
    "country": "Germany"
  },
  "road": "Unter den Linden",
  "house_number": "77",
  "postcode": "10117",
  "city": "Berlin",
  "country": "Germany"
}
```
here, `geoLookup` and `locationString` point to Marwitz, but the main address fields still contain the previously saved Berlin address.

**Comment by Abhinegi2** [comment]
Looks good now

**Review by Abhinegi2 — APPROVED** [review]

File: src/app/features/location/map-popup/map-popup.component.ts [hunk 2]
```diff
@@ -213,17 +215,17 @@ export class MapPopupComponent {
   }
 
   updateLocation(event: GeoLocation | undefined) {
-    let updatedLocation = event;
+    let updatedLocation = enrichGeoLocation(event);
```
Comment (sleidig) [inline src/app/features/location/map-popup/map-popup.component.ts]: can we move the logic of enriching the location to the service, so that different components do not have to call this themselves?
This will make it easier to use from various places and follows the convention to ideally keep logic in services and only UI actions in components.

Or maybe, we could turn the GeoLocation interface into a class that enriches itself on construction/update?

File: src/app/features/location/geo.service.spec.ts [hunk 3]
```diff
@@ -148,6 +155,41 @@ describe("GeoService", () => {
...
+      road: "Village Road",
+      postcode: "99999",
+      village: "Smallville",
+    });
+    const formatted = service.reformatDisplayName(testResult);
+    expect(formatted.display_name).toBe("Village Road, 99999 Smallville");
+    expect(formatted.city).toBe("Smallville");
+  });
+
+  it("should normalize address parts on lookup results for PDF templating", async () => {
```
Comment (sleidig) [inline src/app/features/location/geo.service.spec.ts:169]: while we implement this now for PDF templating, the functionality itself is not tied to it. So let's keep it general

```suggestion
  it("should normalize address parts on lookup results", async () => {
```

File: src/app/features/location/geo.service.ts [hunk 4]
```diff
@@ -24,6 +24,12 @@ import {
...
+  city?: string;
+  country?: string;
```
Comment (sleidig) [inline src/app/features/location/geo.service.ts]: Should these really be added here also? What we populate is on the `GeoLocation` not `GeoResult`, I think? (the two are a bit confusing, GeoResult is what we get from the external API, if I remember correctly, while GeoLocation is what we build from that and save to our DB.

File: src/app/features/location/map-popup/map-popup.component.ts (hunk 2 above)
Comment (sadaf895) [reply src/app/features/location/map-popup/map-popup.component.ts]: I did not convert it to a class because centralizing the enrichment in the GeoService already meets primary goals(single access point, keep logic in services)and converting it into class is a larger API change and will touch more code and tests.

File: src/app/features/location/map-popup/map-popup.component.ts [hunk 5]
```diff
@@ -117,13 +117,15 @@ export class MapPopupComponent {
...
 
-    // Only update geoLookup, keep the user's address string as-is
+    // Only update geoLookup. Do not carry over stale top-level address fields
+    // from the previously selected location — that causes conflicts where
+    // `geoLookup` points to a new place but `road/postcode/city` remain old.
     this.updateLocation({
-      ...this.selectedLocation,
+      locationString: this.selectedLocation?.locationString,
       geoLookup: geoResult,
     });
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/features/location/map-popup/map-popup.component.ts:130]: this is intentional because we ask the user later for confirmation about changing the text
